File: scala/src/main/scala/advent2021/dayTwo.scala
package advent2021
// (depth, horiz, aim)
object DayTwo {
  def partOne(input: Seq[String]) = {
    val opcodes = input.map(_.split(" ")).map { case Array(command, n) =>
      (command, n.toInt)
    }
    var (depth, horiz) = opcodes
      .map { opcode =>
        opcode match {
          case ("forward", n) => (0, n)
          case ("down", n)    => (n, 0)
          case ("up", n)      => (-n, 0)
        }
      }
      .reduce(((x, y) => (x._1 + y._1, x._2 + y._2)))

    depth * horiz
  }

  def partTwo(input: Seq[String]) = {
    val opcodes = input.map(_.split(" ")).map { case Array(command, n) =>
      (command, n.toInt)
    }
    val (depth, horiz, _) = opcodes
      .map { opcode =>
        opcode match {
          case ("forward", n) => (n, n, 0)
          case ("down", n)    => (0, 0, n)
          case ("up", n)      => (0, 0, -n)
        }
      }
      .prepended((0, 0, 0))
      .reduce(((x, y) => (x._1 + (x._3 * y._1), x._2 + y._2, x._3 + y._3)))
    depth * horiz
  }
}

File: advent_ex/lib/day_two.ex
defmodule AdventEx.DayTwo do
  @spec tasks(Enum.t()) :: true
  def tasks(day_input) do
    IO.puts("Day Two, part 1: #{part_one(day_input)}")
    IO.puts("Day Two, part 2: #{part_two(day_input)}")
    true
  end

  @spec part_one(Enum.t()) :: non_neg_integer
  def part_one(day_input) do
    day_input
    |> Stream.map(fn x -> x |> String.split() |> List.to_tuple() end)
    |> Stream.map(fn {op, n} -> {op, n |> Integer.parse() |> elem(0)} end)
    |> Enum.map(fn instruction ->
      case instruction do
        {"forward", n} ->
          {0, n}

        {"up", n} ->
          {-n, 0}

        {"down", n} ->
          {n, 0}
      end
    end)
    |> Enum.reduce(fn {x, y}, {a, b} -> {x + a, y + b} end)
    |> then(fn {x, y} -> x * y end)
  end

  @spec part_two(Enum.t()) :: non_neg_integer
  def part_two(day_input) do
    formatted_input =
      day_input
      |> Stream.map(fn x -> x |> String.split() |> List.to_tuple() end)
      |> Stream.map(fn {op, n} -> {op, n |> Integer.parse() |> elem(0)} end)

    formatted_input
    |> Enum.map(fn instruction ->
      case instruction do
        {"forward", n} ->
          {n, n, 0}

        {"up", n} ->
          {0, 0, -n}

        {"down", n} ->
          {0, 0, n}
      end
    end)
    |> Enum.reduce({0, 0, 0}, fn {dep, hor, aim}, {acc_d, acc_h, acc_a} ->
      vector = acc_a * dep
      {acc_d + vector, acc_h + hor, acc_a + aim}
    end)
    |> then(fn {x, y, _} -> x * y end)
  end
end
